[FEAT] API Data integration

markshenouda opened the pull request:
<img width="1914" height="865" alt="image" src="https://github.com/user-attachments/assets/fc5a2742-02c2-4452-9e84-e285e4f5d87c" />
<img width="1914" height="865" alt="image" src="https://github.com/user-attachments/assets/bd739b71-9e5d-48cc-96d2-b6da597eebc2" />
<img width="1914" height="865" alt="image" src="https://github.com/user-attachments/assets/ab861ad6-36c2-4ee3-8cd7-7160232c2b6e" />
<img width="1914" height="865" alt="image" src="https://github.com/user-attachments/assets/e20750aa-6e58-4ca2-8f58-ffc7187067fb" />

markshenouda committed:
update: Token details page

markshenouda committed:
update: Lane latest data

markshenouda committed:
fix: lint

markshenouda committed:
remove: unused console.log

markshenouda committed:
remove: unused console.log

markshenouda committed:
update: token drawer to include verifiers

Zelig880 commented on src/components/CCIP/Drawer/LaneDrawer.tsx:
```diff
@@ -36,6 +41,27 @@ function LaneDrawer({
+      setIsLoadingRateLimits(false)
+    }
+
+    fetchRateLimits()
```
You are not cleaning up the Fetch Request. This is going to results in many fetch possibly being run at the same time.

Zelig880 commented on src/components/CCIP/Drawer/LaneDrawer.tsx:
```diff
@@ -184,10 +210,29 @@ function LaneDrawer({
+                    // Get standard and FTF rate limits
+                    const allLimits = tokenRateLimits
+                      ? realtimeService.getAllRateLimitsForDirection(tokenRateLimits, direction)
+                      : { standard: null, ftf: null }
```
We are in control of the "getAllRateLimitsForDirection" so we should just return the default or whatever we want to simplify the code below. we should not have to know the "structure" to default it to null here.

Zelig880 commented on src/components/CCIP/Drawer/LaneDrawer.tsx:
```diff
@@ -235,23 +280,68 @@ function LaneDrawer({
+                                }}
+                              />
+                            </span>
+                          )}
```
All these conditions should not happen in the UI. Please abstruct it.. If you find yourself doign an if/ifsele/else, you should abstract. It is very hard to ready

Zelig880 replied:
Same apply to all the i, if else present in this PR

Zelig880 commented on src/components/CCIP/Drawer/TokenDrawer.tsx:
```diff
@@ -53,7 +59,15 @@ function TokenDrawer({
 }) {
   const [search, setSearch] = useState("")
-  const [inOutbound, setInOutbound] = useState<LaneFilter>(LaneFilter.Outbound)
+  const [activeTab, setActiveTab] = useState<"outbound" | "inbound" | "verifiers">("outbound")
```
Not sure why you went away from using a type, if required add a new one, but this does not seem correct

Zelig880 commented on src/components/CCIP/Drawer/TokenDrawer.tsx:
```diff
@@ -65,6 +79,29 @@ function TokenDrawer({
+        const destination = activeTab === "outbound" ? destinationChain : network.key
+        const laneKey = `${source}-${destination}`
+
+        const response = await realtimeService.getLaneSupportedTokens(source, destination, environment)
```
We are using "await" in a for loop. This may slow things down. I wouls suggest to run all this in parallel.

Zelig880 commented on src/components/CCIP/Drawer/TokenDrawer.tsx:
```diff
@@ -65,6 +79,29 @@ function TokenDrawer({
+  // Fetch rate limits for all lanes
+  useEffect(() => {
+    const fetchAllRateLimits = async () => {
+      const realtimeService = new RealtimeDataService()
```
Apply to all useEffect, `const realtimeService = new RealtimeDataService()`, should NOT be initiated within the useEffect, as it iw ill result in logs of memory usage and garbage collection work for no reason.

Zelig880 commented on src/components/CCIP/Drawer/TokenDrawer.tsx:
```diff
@@ -138,141 +175,253 @@ function TokenDrawer({
                 },
               ]}
-              onChange={(key) => setInOutbound(key as LaneFilter)}
+              onChange={(key) => setActiveTab(key as "outbound" | "inbound" | "verifiers")}
```
As mentioned above, please re-introduce the type

Zelig880 commented on src/components/CCIP/Drawer/TokenDrawer.tsx:
```diff
@@ -138,141 +175,253 @@ function TokenDrawer({
+                        verifier.type.toLowerCase().includes(searchLower)
+                      )
+                    })
+                    .map((verifier) => (
```
This operation should not happen here. Make a reusable function for this!

Zelig880 commented on src/components/CCIP/Drawer/TokenDrawer.tsx:
```diff
@@ -138,141 +175,253 @@ function TokenDrawer({
+                      if (!value || value === "0") return "0"
+                      const numValue = BigInt(value)
+                      const formatted = Number(numValue) / 1e18
+                      return formatted.toLocaleString(undefined, { maximumFractionDigits: 2 })
```
This logic should not live in a the UI

Zelig880 replied:
Also, lots of this logic is the same as the other file, please abstract

Zelig880 commented on src/components/CCIP/Drawer/TokenDrawer.tsx:
```diff
@@ -138,141 +175,253 @@ function TokenDrawer({
+                              ? allLimits.standard.isEnabled
+                                ? formatRateLimit(allLimits.standard.capacity)
+                                : "Disabled"
+                              : "N/A"}
```
Other occurances of If/elseif/else

Zelig880 reviewed (changes requested):
One more generic comment, it does not seem like we are handling Failure of the APi request well. We log the issue, but the UI does not seem to have a "error" and may default to a value like 0 while it is incorrect.

markshenouda committed:
refactor: extract realtime data logic into reusable hooks and components

- Create custom hooks for data fetching:
  * useTokenRateLimits - single lane rate limits
  * useMultiLaneRateLimits - multiple lanes in parallel
  * useTokenFinality - token finality data

- Add RateLimitCell component to replace complex nested ternaries

- Create singleton realtimeDataService instance to avoid multiple instantiations

- Add rate-limit-formatter utilities for consistent formatting

- Update LaneDrawer, TokenDrawer, and TokenChainsTable to use new hooks

- Add proper cleanup in useEffect hooks to prevent memory leaks

Benefits:
- Reduced code duplication (~133 lines removed)
- Improved readability with <RateLimitCell> vs 7-level nested ternaries
- Better separation of concerns
- More testable code
- Proper cleanup handling

markshenouda committed:
fix: use enums for tabs

markshenouda committed:
refactor: cleanup Lane's table logic

markshenouda committed:
refactor: token drawer filtering logic

markshenouda committed:
fix: token grid in chain page design

markshenouda committed:
fix: the token card desgin

markshenouda committed:
fix: lint

markshenouda committed:
feat: add verifiers to the search (#184)

* feat: add verifiers to the search

* refactor: simplify verifiers mapping in Chain component

* refactor: simplify verifiers mapping in CCIP components

* refactor: optimize token mapping for verifiers in CCIP components

markshenouda committed:
feat: enhance verifiers data handling with token support

markshenouda committed:
fix: typecheck
